### PR TITLE
Add backup_dvd and HANDBRAKE_LOCAL cfg options.

### PR DESCRIPTION
**Pull request by garbled1**
# Description

Bring back old ARM style of using MakeMKV to copy DVD's to the HD prior
to transcode (with option backup_dvd). Add a HANDBRAKE_LOCAL option. With these two, I can run
ARM semi-distributed via Slurm. It also allows you to do something like:
HANDBRAKE_CLI: ssh hostname HandBrakeCLI
If you just want to do the transcode on a single remote host.

This opens the door to potentially being able to do fully distributed transcoding, for now, it just does 1 disc per remote host. This is basically step 1 in order to do https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/94 I suspect step 2 is to pull in the python slurm library, which may or may not be far too heavyweight.

This is a redo of PR #495

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

Installed slurm_client on host with bluray player. Built 2 VM's with a slurm cluster. Set HANDBRAKE_CLI to "srun --exclusive HandBrakeCLI" and RIPMETHOD to "backup_dvd". Insert 2 discs, and it distributed the transcoding into the cluster. (with generous amounts of NFS sprinkled in).

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [X] Debian 10
- [X] Debian 11
- [ ] Docker
- [ ] Other (Please state here)

# Checklist:

- [X] **I have submitted this PR against the `v2_devel` branch (Unless the main branch has a security issue)**
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have tested that my fix is effective or that my feature works

### CODE REVIEW

**Comment by 1337-server** [comment]
Would you mind submitting job logs for both DVD and Blu-ray ?

Just while we wait for reviews.

**Comment by garbled1** [comment]
It's working on some blurays now, so will try to grab one when it's done.  Here is a dvd.
[ANNIHILATION.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/9173799/ANNIHILATION.log)

The main difference, is you can see the:
srun: job 17643 queued and waiting for resources
srun: job 17643 has been allocated resources
[08:09:45] hb_init: starting libhb thread

Where we run the handbrake via srun, and srun waits for an available node to do the processing.   Right now I have 4 BR's and 5 DVD's in the queue, and this is my squeue:
```
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
             17742       arm HandBrak      arm PD       0:00      1 (Resources)
             17743       arm HandBrak      arm PD       0:00      1 (Priority)
             17744       arm HandBrak      arm PD       0:00      1 (Priority)
             17745       arm recode_t   nobody PD       0:00      1 (Priority)
             17746       arm HandBrak      arm PD       0:00      1 (Priority)
             17747       arm HandBrak      arm PD       0:00      1 (Priority)
             17748       arm HandBrak      arm PD       0:00      1 (Priority)
             17749       arm HandBrak      arm PD       0:00      1 (Priority)
             17738       arm HandBrak      arm  R    1:55:36      1 slurm2
             17741       arm HandBrak      arm  R       1:49      1 slurm1
```
( the nobody one in there, is actually one of my plex transcodes in queue.  I basically queue them into the cluster too).  It basically does one file per disc at a time, and as each one finishes, the next file from the disc is queued.  However, I can just load all the discs in one go, and then just let the remote nodes grind it out.   If I had a power failure or something during this, I'd have to kinda start over, but that's true with normal arm too.
One massive benefit, is now my desktop doesn't slow down at all while I'm ripping,  it basically has no CPU impact locally.

My config is basically:
```
MAX_CONCURRENT_TRANSCODES: 100
RIPMETHOD: "backup_dvd"
HANDBRAKE_CLI: srun --exclusive HandBrakeCLI
HANDBRAKE_LOCAL: HandBrakeCLI
```

**Comment by garbled1** [comment]
[MAN_GOLDEN_GUN.log.gz](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/9174512/MAN_GOLDEN_GUN.log.gz)
Sorry for the gzip, but this one is huge.

**Comment by shitwolfymakes** [comment]
I'm not sure if I fully understand this. How is adding this functionality any different from setting the `RIPMETHOD` to `mkv`?

Love the idea of the distributed encoding!

**Comment by 1337-server** [comment]
> I'm not sure if I fully understand this. How is adding this functionality any different from setting the `RIPMETHOD` to `mkv`?

They explain this here → https://github.com/automatic-ripping-machine/automatic-ripping-machine/pull/495#issuecomment-1118523395

**Comment by garbled1** [comment]
> I'm not sure if I fully understand this. How is adding this functionality any different from setting the `RIPMETHOD` to `mkv`?

I explain it better in the comments for #495  
Basically, I want the best of both worlds.  I want BR's pulled as backups, and DVD's saved to disk as mkv's.  This allows me to pull the full disc down to an NFS mount regardless of the type, and then do all the transcoding from another node, and not be subject to only doing mainfeature or other restrictions.
 
> Love the idea of the distributed encoding!

It actually works very very well when you do multiple discs at once.  On my 2-node cluster, if you just do one disc, you get basically remote encoding on the first node of the cluster, but with multiple discs, it interleaves all of the transcodes onto the 2 nodes.  It's almost fascinating how I can load in 6 DVD's and 3 BR's and they all finish pretty much at the same time.

**Comment by shitwolfymakes** [comment]
Huh. In that case we should put it on the roadmap to just have separate ripmethods for each disc type. This will make a great stopgap in the meantime.

**Comment by shitwolfymakes** [comment]
Where can I find the docker container for this?

**Comment by 1337-server** [comment]
> Where can I find the docker container for this?

You would need to pull the PR and build it

File: setup/arm.yaml [hunk 1]
```diff
@@ -176,11 +176,12 @@ CHOWN_GROUP:
 ########################
 
 # Method of MakeMKV to use for Blu-ray discs
-# Options are "mkv" or "backup".
+# Options are "mkv", "backup", or "backup_dvd".
 # backup decrypts the dvd and then copies it to the hard drive.
 # This allows HandBrake to apply some of its analytical abilities such as the main-feature identification.
 # This method seems to offer success on Blu-ray discs that fail in "mkv" mode.
-# *** NOTE: MakeMKV only supports the backup method on BluRay discs.
+# *** NOTE: MakeMKV only supports the backup or backup_dvd method on BluRay discs.
```
[the remark this answers is not in the document]
Comment (garbled1) [reply setup/arm.yaml:183]: I'm not sure what to correct it to?  It used to just say only "backup".

File: arm/ripper/makemkv.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def makemkv(logfile, job):
     rawpath = setup_rawpath(job, os.path.join(str(job.config.RAW_PATH), str(job.title)))
     logging.info(f"Processing files to: {rawpath}")
     # Rip bluray
-    if job.config.RIPMETHOD == "backup" and job.disctype == "bluray":
+    if (job.config.RIPMETHOD == "backup" or job.config.RIPMETHOD == "backup_dvd") and job.disctype == "bluray":
```
[the remark this answers is not in the document]
Comment (garbled1) [reply arm/ripper/makemkv.py:55]: It's not only for dvd's.  It modifies both dvd and bluray behaviour.

File: setup/arm.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shitwolfymakes) [reply setup/arm.yaml:183]: `backup_dvd` does not apply to Blu-ray discs, as it is a dvd-only option

File: arm/ripper/makemkv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shitwolfymakes) [reply arm/ripper/makemkv.py:55]: gotcha

File: arm/ui/comments.json [hunk 3]
```diff
@@ -48,13 +48,15 @@
   "SET_MEDIA_OWNER": "# Don't bother. This does not do anything yet",
   "CHOWN_USER": "",
   "CHOWN_GROUP": "",
-  "RIPMETHOD": "# Method of MakeMKV to use for Blu Ray discs.  Options are \"mkv\" or \"backup\".\n# backup decrypts the dvd and then copies it to the hard drive.  This allows HandBrake to apply some of it's\n# analytical abilities such as the main-feature identification.  This method seems to offer success on bluray \n# discs that fail in \"mkv\" mode. *** NOTE: MakeMKV only supports the backup method on BluRay discs.  ARM does\n# not use MakeMKV for DVDs.   ",
+  "RIPMETHOD": "# Method of MakeMKV to use for Blu Ray discs.  Options are \"mkv\", \"backup\" or \"backup_dvd\".\n# backup decrypts the dvd and then copies it to the hard drive.  This allows HandBrake to apply some of it's\n# analytical abilities such as the main-feature identification.  This method seems to offer success on bluray \n# discs that fail in \"mkv\" mode. *** NOTE: MakeMKV only supports the backup or backup_dvd method on BluRay discs.\n# backup_dvd forces arm to extract the dvd with MakeMKV prior to the Handbrak
```
[the remark this answers is not in the document]
Comment (shitwolfymakes) [reply arm/ui/comments.json:51]: backup_dvd does not apply to Blu-ray discs, as it is a dvd-only option

File: arm/ui/comments.json [hunk 4]
```diff
@@ -48,13 +48,15 @@
...
+e step",
```
Comment (shitwolfymakes) [inline arm/ui/comments.json:52]: This newline breaks the json file, which in turn breaks the ability to save ripper settings.

**Review by shitwolfymakes — CHANGES_REQUESTED** [review]

**Comment by shitwolfymakes** [comment]
After fixing the broken json and rebuilding the docker container, I am still getting a failed rip due to `json_api.process_handbrake_logfile` erroring `Can't find index`.

I use this Blu-ray disc for testing so I know it's good. The RIPMETHOD is set to `mkv`. Attached is the logfile.
[MOULIN_ROUGE_165936741511.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/9236220/MOULIN_ROUGE_165936741511.log)

**Comment by 1337-server** [comment]
> `json_api.process_handbrake_logfile` erroring `Can't find index`.

Those should be soft errors. But it looks like that ARM isn't correctly logging the errors to the log file. Don't think it's related to this issue, but ill create a new issue for it.

**Comment by shitwolfymakes** [comment]
Upon looking at it, the output is looking similar to the issue in #546

**Comment by 1337-server** [comment]
> Upon looking at it, the output is looking similar to the issue in #546

Can you share the log ?

File: arm/ui/comments.json (hunk 4 above)
Comment (garbled1) [reply arm/ui/comments.json]: fixed

File: arm/ui/comments.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (garbled1) [reply arm/ui/comments.json]: perhaps I named it poorly, but the idea is that it's the "backup" option, that also works on dvd's.  So it is intended to modify the behavior of both.

**Review by shitwolfymakes — APPROVED** [review]

**Review by 1337-server — APPROVED** [review]

File: arm/ui/comments.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (shitwolfymakes) [reply arm/ui/comments.json]: Gotcha

**Comment by shitwolfymakes** [comment]
My test of the docker image didn't show any issues introduced

**Comment by garbled1** [comment]
Thank you!

**Comment by 1337-server** [comment]
> Thank you!

Thanks for contributing! Even more so with having to redo the PR.

**Pull request merged**